app/scripts/lib/AbstractPetnamesBridge.ts

```diff
@@ -0,0 +1,240 @@
+      );
+    }
+    if (this.#synchronizingDirection !== 'NONE') {
+      return; // Ignore calls while updating in the opposite direction
+    }
+
+    this.#synchronizingDirection = direction;
+
+    const [newEntries, prevEntries] =
+      direction === 'Source->Petnames'
+        ? [this.getSourceEntries(), this.#getPetnameEntries()]
+        : [this.#getPetnameEntries(), this.getSourceEntries()];
+
+    const changeList = this.#computeChangeList(prevEntries, newEntries);
+    this.#applyChangeList(changeList);
+
+    this.#synchronizingDirection = 'NONE';
+  }
+
+  /**
+   * Extract PetnameEntry objects from the name controller state.
+   */
+  #getPetnameEntries(): PetnameEntry[] {
+    const { names } = this.#nameController.state;
+    const entries: PetnameEntry[] = [];
+    for (const type of Object.values(NameType)) {
+      for (const variation of Object.keys(names[type])) {
+        for (const value of Object.keys(names[type][variation])) {
+          const entry = names[type][variation][value];
+          entries.push({
+            value,
+            type,
+            name: entry.name,
+            sourceId: entry.sourceId,
+            variation,
+          } as SetNameRequest);
+        }
+      }
+    }
+    return entries;
+  }
+
+  /**
+   * Updates Petnames with the given entry.
+   *
+   * @param type - The type of change that occurred.
+   * @param entry - The entry to update the name controller with.
+   */
+  #updatePetnameEntry(type: ChangeType, entry: PetnameEntry): void {
+    if (type === ChangeType.DELETED) {
+      this.#nameController.setName({ ...entry, name: null });
+    } else {
+      // ADDED or UPDATED
+      this.#nameController.setName(entry);
+    }
+  }
+
+  /**
+   * Get a string key for the given entry.
+   *
+   * @param entry
+   */
+  #getKey(entry: PetnameEntry): string {
+    return `${entry.type}/${entry.variation}/${entry.value}`;
+  }
+
+  /**
+   * Computes the list of changes between the previous and new entries.
+   *
+   * @param prevEntries - The previous entries.
+   * @param newEntries - The new entries.
+   * @returns A ChangeList object representing the changes that occurred between prevEntries and newEntries.
+   */
+  #computeChangeList(
+    prevEntries: PetnameEntry[],
+    newEntries: PetnameEntry[],
+  ): ChangeList {
+    const added: PetnameEntry[] = [];
+    const updated: PetnameEntry[] = [];
+    const deleted: PetnameEntry[] = [];
+
+    const prevEntriesMap = new Map(
+      prevEntries.map((e) => [this.#getKey(e), e]),
+    );
+    const newEntriesMap = new Map(newEntries.map((e) => [this.#getKey(e), e]));
+
+    newEntriesMap.forEach((newEntry, newKey) => {
+      const oldEntry = prevEntriesMap.get(newKey);
+      if (oldEntry) {
+        if (newEntry.name !== oldEntry.name) {
+          updated.push(newEntry);
+        }
+      } else {
+        added.push(newEntry);
+      }
+    });
+
+    if (!this.#isTwoWay) {
```

Comment:
Do we now want the inverse here?

As we want address book removals (two way) to be removed from petnames and vice versa, but for account identities (one way) we don't want anything never in the account identities to be removed from petnames?

Reply:
Yep, good catch. Done in next update.